specs/core/1_shard-data-chains.md

```diff
@@ -299,78 +331,201 @@ def compute_crosslink_data_root(blocks: Sequence[ShardBlock]) -> Hash:
+        REWARD_COEFFICIENT_BASE * BASE_REWARD_FACTOR // integer_squareroot(total_balance) // BASE_REWARDS_PER_EPOCH
+    )
 
     # Check attestations
-    attester_committee = get_persistent_committee(beacon_state, shard, block.core.slot)
+    attester_committee = get_persistent_committee(beacon_state, state.shard, block.core.slot)
     pubkeys = []
+    attestations = 0
+
     for i, index in enumerate(attester_committee):
         if block.core.attester_bitfield[i]:
             pubkeys.append(beacon_state.validators[index].pubkey)
-    for i in range(len(attester_committee), MAX_PERSISTENT_COMMITTEE_SIZE * 2):
-        assert block.attester_bitfield[i] is False
+            add_reward(state, beacon_state, index, base_reward)
+            attestations += 1
+
+    for i in range(len(attester_committee), MAX_PERSISTENT_COMMITTEE_SIZE):
+        assert block.core.attester_bitfield[i] is False or block.core.attester_bitfield[i] == 0  # TODO: FIX Bitvector
+
     assert bls_verify(
         pubkey=bls_aggregate_pubkeys(pubkeys),
-        message_hash=candidate.core.parent_root,
-        signature=candidate.signatures.attestation_signature,
-        domain=get_domain(beacon_state, DOMAIN_SHARD_ATTESTER, compute_epoch_of_shard_slot(candidate.core.slot))
+        message_hash=block.core.parent_root,
+        signature=block.signatures.attestation_signature,
+        domain=get_domain(beacon_state, DOMAIN_SHARD_ATTESTER, compute_epoch_of_shard_slot(block.core.slot))
     )
 
     # Check proposer
-    proposer_index = get_shard_block_proposer_index(beacon_state, shard, candidate.core.slot)
+    proposer_index = get_shard_block_proposer_index(beacon_state, state.shard, block.core.slot)
     assert proposer_index is not None
+    add_reward(state, beacon_state, proposer_index, Gwei(attestations * base_reward // PROPOSER_REWARD_QUOTIENT))
     assert bls_verify(
         pubkey=beacon_state.validators[proposer_index].pubkey,
-        message_hash=hash_tree_root(candidate.core),
-        signature=candidate.signatures.proposer_signature,
-        domain=get_domain(beacon_state, DOMAIN_SHARD_PROPOSER, compute_epoch_of_shard_slot(candidate.core.slot)),
+        message_hash=hash_tree_root(block.core),
+        signature=block.signatures.proposer_signature,
+        domain=get_domain(beacon_state, DOMAIN_SHARD_PROPOSER, compute_epoch_of_shard_slot(block.core.slot)),
     )
 
-    return True
+    # Process and update block data fees
+    add_fee(state, beacon_state, proposer_index, Gwei(state.basefee * len(block.core.data) // SHARD_BLOCK_SIZE_LIMIT))
+    QUOTIENT = SHARD_BLOCK_SIZE_LIMIT * BASEFEE_ADJUSTMENT_FACTOR
+    if len(block.core.data) > SHARD_BLOCK_SIZE_TARGET:
+        state.basefee += Gwei(min(1, state.basefee * (len(block.core.data) - SHARD_BLOCK_SIZE_TARGET) // QUOTIENT))
+    elif len(block.core.data) < SHARD_BLOCK_SIZE_TARGET:
+        state.basefee -= Gwei(min(1, state.basefee * (len(block.core.data) - SHARD_BLOCK_SIZE_TARGET) // QUOTIENT))
```

Comment:
It's possible that the result `state.basefee` is negative.
Do you mean:
```suggestion
        state.basefee = Gwei(min(1, state.basefee - state.basefee * (len(block.core.data) - SHARD_BLOCK_SIZE_TARGET) // QUOTIENT))
```

Reply:
Ah, I made one mistake: it should be max instead of min. The change definitely can't be bigger than the current basefee (if it can, that's a bug). So I think it's fine as is. The goal of the max is to ensure that under any conditions the fee can adjust by at least 1 per block.